java-checks/src/main/java/org/sonar/java/checks/ModifiersOrderCheck.java

```diff
@@ -65,23 +66,45 @@ public void visitNode(Tree tree) {
   }
 
   private static ModifierTree getFirstBadlyOrdered(ModifiersTree modifiersTree) {
-    int modifierIndex = -1;
+    ListIterator<ModifierTree> modifiersIterator = modifiersTree.listIterator();
+    skipAnnotations(modifiersIterator);
     Modifier[] modifiers = Modifier.values();
-    for (ModifierTree modifier : modifiersTree) {
+    int modifierIndex = 0;
+    while (modifiersIterator.hasNext()){
+      ModifierTree modifier = modifiersIterator.next();
       if (modifier.is(Kind.ANNOTATION)) {
-        if (modifierIndex >= 0) {
-          return modifier;
-        }
-      } else {
-        if (modifierIndex < 0) {
-          modifierIndex = 0;
-        }
-        ModifierKeywordTree mkt = (ModifierKeywordTree) modifier;
-        for (; modifierIndex < modifiers.length && !mkt.modifier().equals(modifiers[modifierIndex]); modifierIndex++) {
-          // We're just interested in the final value of modifierIndex
-        }
-        if (modifierIndex == modifiers.length) {
-          return modifier;
+        break;
+      }
+      ModifierKeywordTree mkt = (ModifierKeywordTree) modifier;
+      for (; modifierIndex < modifiers.length && !mkt.modifier().equals(modifiers[modifierIndex]); modifierIndex++) {
+        // We're just interested in the final value of modifierIndex
+      }
+      if (modifierIndex == modifiers.length) {
+        return modifier;
+      }
+    }
+    return testOnlyAnnotationsAreLeft(modifiersIterator);
+  }
+
+  /**
+   * Move iterator on the first element which is not an annotation
+   */
+  private static void skipAnnotations(ListIterator<ModifierTree> modifiersIterator) {
+    while (modifiersIterator.hasNext() && modifiersIterator.next().is(Kind.ANNOTATION)) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Either remove or fill this block of code. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00108)